adagrams/game.py

```diff
@@ -1,11 +1,159 @@
+    'W': 2, 
+    'X': 1, 
+    'Y': 2, 
+    'Z': 1
+}
+
+
 def draw_letters():
-    pass
+
+    my_list = [] #creating an empty list to collect randomly picked letter
+
+    #converting dictionary keys into list representing the 'count' times of letter and storing in pool_list
+    pool_list = []
+    for letter, count in LETTER_POOL.items():
+        pool_list += [letter] * count # add letter 'count' times to pool_list
+        
+    while len(my_list) < 10:
+        random_key = random.randint(0, len(pool_list) - 1) #randomly select index from pool_list
+        
+        random_letter = pool_list[random_key] #getting letter associated with index
+        pool_list.remove(random_letter) #removing letter from pool_list
+        my_list.append(random_letter) #adding this letter to my_list
+
+    return my_list
+
+
+letter_bank = draw_letters() # retreiving my_list of 10 listter from draw_letter function
+
+word = "SANTAAA"
+
 
 def uses_available_letters(word, letter_bank):
-    pass
+
+    # getting user input as string 
+    upper_word = word.upper()
+
+    letter_bank_dict = {} #empty dictionary to store 
+    inputed_word_dict = {} #empty dictionary to store letter and frequency from user
+
+
+    # adding upper_word into inputed_word_dict with letter: count format
+    for letter in upper_word:
+        if letter not in inputed_word_dict:
+            inputed_word_dict[letter] = 1 #Initialize with one if letter added first time
+        else:
+            inputed_word_dict[letter] += 1 #Increment by 1 if letter added additionally
+    
+
+    # converting answer_list to a letter_bank_dict with a format of letter:count
+    for letter in letter_bank:
+        if letter not in letter_bank_dict:
+            letter_bank_dict[letter] = 1
+        else:
+            letter_bank_dict[letter] += 1
+
+
+    # comparing letter_bank_dict and inputed_word_dict 
+    for key, value in inputed_word_dict.items():
+        if key not in letter_bank_dict or value > letter_bank_dict[key]:  # checking if key from inputed_word_dict is in letter_bank_dict and checking if value of inputed_word_dict is greater than value in letter_bank_dict
+            return False
+        
+    return True
+
+
+x = uses_available_letters(word, letter_bank)
+
 
 def score_word(word):
-    pass
+    score_dict = {
+        'A': 1, 'E': 1, 'I': 1, 'O': 1, 'U': 1,
+        'L': 1, 'N': 1, 'R': 1, 'S': 1, 'T': 1,
+        'D': 2, 'G': 2,
+        'B': 3, 'C': 3, 'M': 3, 'P': 3,
+        'F': 4, 'H': 4, 'V': 4, 'W': 4, 'Y': 4,
+        'K': 5,
+        'J': 8, 'X': 8,
+        'Q': 10, 'Z': 10
+    }
+    #getting user input as string 
+    upper_word = word.upper()
+    score = 0
+
+    # incrementing score with values from score_dict
+    for letter in upper_word:
+        if letter in score_dict:
+            score += score_dict[letter]
+    # If the length of the word is 7, 8, 9, or 10: score gets incremented by 8
+    if len(upper_word) in (7,8,9,10):
+        score += 8
```

Comment:
We'd typically see this written like 
```python
if 7 <= len(word) <= 10:
    score += 8
```

We'll cover time complexity in tomorrow's roundtable but the `in` operator has linear time complexity which is higher than the constant time complexity it takes for Python to check a comparison like I have in the code example above. 

Also, since it's a comparison we'd tend to use comparison operators instead of `in` operators.

Reply:
The above suggestion to check the length of the word against 7 and 10 could be written in a way that is even more self-documenting by introducing some constant variables. This does make your code longer but is worth doing so because it makes your code easier to understand

```python
BONUS_LENGTH_MIN = 7
BONUS_LENGTH_MAX = 10
BONUS_POINTS = 8


if BONUS_LENGTH_MIN <= len(word) <= BONUS_LENGTH_MAX:
        total_score = total_score + BONUS_POINTS
```